File: hexagonal-app-load-testing/app-vertx-java/src/main/java/app/api/CpuHandler.java
package app.api;

import app.dto.WorkCpuRequest;
import app.dto.WorkCpuResponse;
import business.service.CpuBoundUseCase;
import business.service.dto.WorkCpuCommand;
import io.netty.handler.codec.http.HttpResponseStatus;
import io.vertx.core.Future;
import io.vertx.ext.web.RoutingContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.function.Function;

public class CpuHandler implements Function<RoutingContext, Future<WorkCpuResponse>> {

    private static final Logger log = LoggerFactory.getLogger(CpuHandler.class);

    private final CpuBoundUseCase cpuBoundUseCase;

    public CpuHandler(CpuBoundUseCase cpuBoundUseCase) {
        this.cpuBoundUseCase = cpuBoundUseCase;
    }

    @Override
    public Future<WorkCpuResponse> apply(RoutingContext routingContext) {

        var workCpuRequest = routingContext.getBodyAsJson().mapTo(WorkCpuRequest.class);
        if (workCpuRequest == null || workCpuRequest.getInputs() == null || workCpuRequest.getInputs().isEmpty()) {

            routingContext.response()
                    .setStatusCode(HttpResponseStatus.BAD_REQUEST.code())
                    .end("CPU work inputs cannot be empty!");
            return Future.succeededFuture();
        }

        log.debug("CPU task with {} inputs", workCpuRequest.getInputs().size());
        var durationInNanos = cpuBoundUseCase.workCpu(new WorkCpuCommand(workCpuRequest.getInputs()));

        routingContext.response().setStatusCode(HttpResponseStatus.ACCEPTED.code());
        return Future.succeededFuture(new WorkCpuResponse(durationInNanos));
    }
}

File: hexagonal-app-load-testing/app-ktor-kotlin/src/main/kotlin/app/api/CpuHandler.kt
package app.api

import WorkCpuRequest
import WorkCpuResponse
import app.DependencyFactory
import business.dto.CpuWorkInputResponse
import business.entity.CpuTask
import business.service.CpuBoundUseCase
import io.ktor.application.*
import io.ktor.features.*
import io.ktor.http.*
import io.ktor.request.*
import io.ktor.response.*
import io.ktor.routing.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.slf4j.LoggerFactory

private val log = LoggerFactory.getLogger("app.CpuHandler")

internal fun Routing.cpuHandler() {

    val cpuBoundUseCase: CpuBoundUseCase = DependencyFactory.cpuBoundUseCase()

    post("/cpu") {

        val requestBody = call.receive(WorkCpuRequest::class)
        if (requestBody.inputs.isEmpty()) {
            throw BadRequestException("CPU work inputs cannot be empty!")
        }

        withContext(Dispatchers.Unconfined) {

            log.debug("CPU task with {} inputs", requestBody.inputs.size)

            val task = CpuTask.Builder(requestBody.inputs).build()
            val responses = cpuBoundUseCase.compute(task)
            val durationInNanos = responses.stream().mapToLong { it: CpuWorkInputResponse -> it.inputProcessingDuration }.sum()
            call.respond(HttpStatusCode.Accepted, WorkCpuResponse(durationInNanos))
        }
    }
}